PYME/cluster/HTTPDataServer.py

```diff
@@ -679,6 +681,64 @@ def get_h5_part(self, path):
         except IOError:
             self.send_error(404, "File not found - %s, [%s]" % (self.path, path))
             return None
+
+    def get_tabular_part(self, path):
+        """
+
+        Parameters
+        ----------
+        path: str
+            path to an hdf or h5r file on the dataserver computer. Specific parts of the file can be accessed by
+            appending, e.g. /Metadata, for a resulting `path` of something like folder/test.h5r/Metadata. The 'part' of
+            the file returned defaults to FitResults if unspecified, and can additionally include a slice to take before
+            returning. In other words, a `path` of 'folder/test.h5r/FitResultsslice(0,100,None)' will return the first
+            hundred elements of the FitResults table in test.h5r.
+
+        Returns
+        -------
+        f: BytesIO
+            Requested part of the file encoded as bytes
+
+        """
+        from PYME.IO.tabular import NestedRecArraySource
+        from PYME.IO import h5rFile
+        ext = '.h5r' if '.h5r' in path else '.hdf'
+        try:
+            filename, part = path.split(ext)
+            ctype = self.guess_type(filename + ext)
+            part = part.lstrip('/').lstrip('\\')
+
+            with h5rFile.openH5R(filename + ext) as h5f:
+                if part == 'Metadata':
+                    mdh = h5f.mdh.to_JSON()
+                    f, length = self._string_to_file(mdh)
+                elif part == 'Events':
+                    # events = h5f.events
+                    #  fixme / make events class with to_JSON method
+                    self.send_error(400, 'serving events from tabular files not yet supported')
+                    return
+                else:
+                    if 'slice' in part:
+                        # convert the slice back from str, avoiding eval for security
+                        part, return_slice = part.strip(')').split('slice(')
+                        return_slice = slice(*(None if s == 'None' else int(s) for s in return_slice.split(',')))
+                    else:
+                        return_slice = slice(None)
+                        if part == '':
+                            part = 'FitResults'
+                    table = h5f.getTableData(part, return_slice)
+                    table = NestedRecArraySource(table, translate_pipeline_keys=True)
+                    f, length = self._string_to_file(table.to_json())
```

Comment:
I suspect we'll run into performance issues serialising as json. Look at what we use to write data to files in the HTTPDataServer `__aggregate` endpoints and/or in `PYME.IO.clusterResults`

Ultimately we serialise as pickle (which has security issues). Using numpys disk format would be better from a security perspective, but significantly slower (although still not nearly as bad as json) need to watch recent numpys in case this is fixed).

Reply:
We should probably refactor `clusterResults.fileResults()` to first call a `clusterResults.format_results()` function which contains the bulk of `fileResults()` and  then the URI translation and `fileFormattedResults()` so that we can re-use `clusterResults.format_results()` when returning tabular parts.  I'll do that now.

Reply:
Overall I think this is a really clean approach, but casting nested structured arrays to json using `format_results` gives us rubbish on nested arrays. This is essentially why I added the `NestedRecArraySource`, but maybe instead we could just add an unnest recarray function somewhere?

Reply:
alternatively, maybe we don't care enough to pass table parts as json and we just stick with .npy